EVENTS.md

```diff
@@ -0,0 +1,416 @@
+    sdk.NewAttribute("action", "promote"),
+    sdk.NewAttribute("batch_id", "6"),
+    sdk.NewAttribute("address", "cosmos1234567"),
+    sdk.NewAttribute("address", "cosmos1765432"),
+    sdk.NewAttribute("batch_id", "7"),
+    sdk.NewAttribute("address", "cosmos19875632"),
+)
+```
+
+### Validation Rules
+
+While the `wasm` and `wasm-*` namespacing does sandbox the smart contract events and limits malicious activity they could
+undertake, we also perform a number of further validation checks on the contracts:
+
+* No attribute key may start with `_`. This is currently used for `_contract_address` and `_code_id` and is reserved for a 
+  namespace for injecting more *trusted* attributes from the `x/wasm` module as opposed to the contract itself
+* Event types are trimmed of whitespace, and must have at least two characters prior to prepending `wasm-`. If the contract returns
+  "  hello\n", the event type will look like `wasm-hello`. If it emits "  a  ", this will be rejected with an error (aborting execution!)
+* Attribute keys and values (both in `attributes` and under `events`) are trimmed of leading/trailing whitespace. If they are empty after
+  trimming, they are rejected as above (aborting the execution). Otherwise, they are passed verbatim.
+
+## Event Details for wasmd
+
+Beyond the basic Event system and emitted events, we must handle more advanced cases in `x/wasm`
+and thus add some more logic to the event processing. Remember that CosmWasm contracts dispatch other
+messages themselves, so far from the flattened event structure, or even a list of list (separated
+by message index in the tx), we actually have a tree of messages, each with their own events. And we must
+flatten that in a meaningful way to feed it into the event system.
+
+Furthermore, with the sub-message reply handlers, we end up with eg. "Contract A execute", "Contract B execute",
+"Contract A reply". If we return all events by all of these, we may end up with many repeated event types and
+a confusing results. However, we may use the standard "message" events to separate the sub-messages as it marks
+where the next one starts. With careful analysis of the "sender" field on these "message" markers, we may be able
+to reconstruct much of the tree execution path. We should ensure all this information is exposed in the most
+consistent way possible.
+
+### Combining Events from Sub-Messages
+
+Each time a contract is executed, it not only returns the `message` event from its call, the `execute` event for the
+contact and the `wasm` event with any custom fields from the contract itself. It will also return the same set of information
+for all messages that it returned, which were later dispatched.
+
+Until 0.18.0, we stripped out the `message` events that were returned when dispatching `Response.messages`. However,
+they contain useful information to trace the topography of the call, especially when we call into native modules,
```

Comment:
If all sdk modules would follow the `distribution` event design, I would totally agree on this approach. Unfortunately the **bank** module emits a `message` event in the keeper. Probably others as well. 
We call the keeper when we transfer coins to the contract account. This would result into 2 different message events where only the wasm one is correct.
A more consistent approach was the message filtering so that we only track the message initiating the whole process via event. We do loose information going this way, but I would prefer this rather than emitting false positives.
We could also address the bank message issue by filtering out the bank message in our `BankCoinTransferrer`. For native modules calling the bank keeper this would not work though.

Reply:
Afaik, we call the router for most dispatching (except ibc) and thus hit the message server as well.

I also hear your point about the non standard issue. But it is very useful to have a separator between the different sub messages.

How about this:
1. Filter out the inconsistent message event emitted by the module.
2. Always prepend a standard event type on the beginning of each submessage. We can define what it is, it is created by x/wasm and will be very consistent.